Cuda refactor, multi GPU support

JohannesGaessler opened the pull request:
This PR is quite large. Its primary goal is to lay the groundwork for the implementation of further CUDA kernels for ggml operations. I am also adding multi GPU support because it's easier to integrate now than it would be at a later point. 

### For Users

Build instructions (Linux):

```
git clone https://github.com/JohannesGaessler/llama.cpp llama.cpp-johannesgaessler
cd llama.cpp-johannesgaessler                               
git fetch
git switch cuda-refactor-8
make LLAMA_CUBLAS=1
```

When compiling with `LLAMA_CUBLAS=1` the program automatically detects the available NVIDIA devices and splits weights proportional to VRAM. There is not yet a CLI argument for setting the tensor splits. The performance increase on my test systems is relatively low (+70% t/s when going from 1x GTX TITAN X to 4x GTX TITAN X). It's possible that there is still a bug that hampers performance. Please do tell me how well (if at all) it works for you. In any case, this PR should already allow you to pool the VRAM of multiple GPUs to load larger models.

### For Developers

~~**This PR is still very much WIP. I will do a refactor to remove artifacts from bad/obsolete design decisions. You can already review the code if you want but many of the flaws are still subject to change.**~~ Should be good now.

On master there are separate functions for invoking CUDA kernels. Apart from invoking the actual CUDA kernels they do other things such as copying data between host and device. This PR adds a template `ggml_cuda_op` that manages

1. the transfer of data between host and device,
2. the dequantization of `src0` (needed for cuBLAS matrix multiplication),
3. the broadcasting of `src1` across `src0` (needed for multiplication),
4. and multi GPU things.

The actual operations now only need to define how the data should be manipulated.

This PR also moves the entry point for invoking CUDA kernels from the ggml function such as `ggml_compute_forward_mul_mat_q_f32` and instead adds a function `ggml_cuda_compute_forward` that is called from `ggml_compute_forward`. For this to work I moved `ggml_task_type` and `ggml_compute_params` from `ggml.c` to `ggml.h`.

This PR adds an int for the layer, an int for the device id, and dedicated device data pointers to `ggml_tensor`. I need these for bookkeeping. I also changed the backends from `GGML_BACKEND_CUDA` and `GGML_BACKEND_OPENCL` to `GGML_BACKEND_GPU` (tensor data on 1 GPU) and `GGML_BACKEND_GPU_SPLIT` (tensor data split across all GPUs). Since I think that we don't want to support the simultaneous use of CUDA and OpenCL it's simpler to just use the same backend types for both implementations and to differentiate via defines.

JohannesGaessler committed:
CUDA op template

JohannesGaessler committed:
ggml_cuda_compute_forward

JohannesGaessler committed:
Tensor parallelism

ENjoyBlue2021 commented:
Thanks so much for adding multi GPU support, I was looking forward to it. You are the king man.
This is extremly useful to increase my vram for 30b.
I'm posting my stats with my 1080ti+1080.
Amazing stuff, its noticeable faster. 

Stats with this version:
```
./main -m '/media/w/PhoenixSSD/oobabooga/text-generation-webui/models/supercot30b-ggml/ggml-model-q5_1.bin' -n 128 --n-gpu-layers 39 --threads 6 --no-mmap -s 1685192470
WARNING: when using cuBLAS generation results are NOT guaranteed to be reproducible.
main: build = 596 (428c342)
main: seed  = 1685192470
ggml_init_cublas: found 2 CUDA devices:
  1. NVIDIA GeForce GTX 1080 Ti
  2. NVIDIA GeForce GTX 1080
llama.cpp: loading model from /media/w/PhoenixSSD/oobabooga/text-generation-webui/models/supercot30b-ggml/ggml-model-q5_1.bin
llama_model_load_internal: format     = ggjt v2 (pre #1508)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 9 (mostly Q5_1)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size = 23269.21 MB
llama_model_load_internal: mem required  = 10646.42 MB (+ 3124.00 MB per state)
llama_model_load_internal: [cublas] offloading 39 layers to GPU
llama_model_load_internal: [cublas] total VRAM used: 14926 MB
....................................................................................................
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 6 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 128, n_keep = 0


 // Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;

namespace MSBuild.ExtensionPack.Platform
{
    /// <summary>
    /// Adds support for a target to find the first non-empty environment variable, by name
llama_print_timings:        load time = 14079.62 ms
llama_print_timings:      sample time =    55.40 ms /   128 runs   (    0.43 ms per token)
llama_print_timings: prompt eval time =   722.20 ms /     2 tokens (  361.10 ms per token)
llama_print_timings:        eval time = 54367.11 ms /   127 runs   (  428.09 ms per token)
llama_print_timings:       total time = 68522.85 ms

```

Compare old Version:
```
./main -m '/media/w/PhoenixSSD/oobabooga/text-generation-webui/models/supercot30b-ggml/ggml-model-q5_1.bin' -n 128 --n-gpu-layers 17 --threads 6 --no-mmap -s 1685192470
WARNING: when using cuBLAS generation results are NOT guaranteed to be reproducible.
main: build = 583 (7e4ea5b)
main: seed  = 1685192470
llama.cpp: loading model from /media/w/PhoenixSSD/oobabooga/text-generation-webui/models/supercot30b-ggml/ggml-model-q5_1.bin
llama_model_load_internal: format     = ggjt v2 (pre #1508)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 9 (mostly Q5_1)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size = 23269.14 MB
llama_model_load_internal: mem required  = 19066.59 MB (+ 3124.00 MB per state)
llama_model_load_internal: [cublas] offloading 17 layers to GPU
llama_model_load_internal: [cublas] total VRAM used: 6506 MB
....................................................................................................
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 6 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 128, n_keep = 0


 // Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;

namespace MSBuild.ExtensionPack.Platform
{
    /// <summary>
    /// Adds support for a target to find the first non-empty environment variable, by name
llama_print_timings:        load time = 11117.69 ms
llama_print_timings:      sample time =    55.35 ms /   128 runs   (    0.43 ms per token)
llama_print_timings: prompt eval time =   896.98 ms /     2 tokens (  448.49 ms per token)
llama_print_timings:        eval time = 82108.70 ms /   127 runs   (  646.53 ms per token)
llama_print_timings:       total time = 93302.20 ms

```

SlyEcho commented:
Would be too much to ask cross platform (Nvidia + AMD) support LMAO.

I will try to check later, I think I can access a machine with 2x 2080 Ti

JohannesGaessler commented:
Performance numbers from my test machine with an i5-4570S, 16 GB of RAM @ 1600 MHz, and a GTX 1070 + a GTX 1050 ti:

| Model    | GPU                    |     ms/t |      t/s |
|----------|------------------------|----------|----------|
| 7b q4_0  | GTX 1070               |    71.37 |    14.01 |
| 7b q4_0  | GTX 1070 + GTX 1050 ti |    68.66 |    14.56 |
| 7b q4_0  | GTX 1070               |    71.37 |    14.01 |
| 13b q4_0 | GTX 1070               |   134.19 |     7.45 |
| 13b q4_0 | GTX 1070 + GTX 1050 ti |   128.13 |     7.80 |
| 33b q4_0 | GTX 1070               | Unusable | Unusable |
| 33b q4_0 | GTX 1070 + GTX 1050 ti |   575.12 |     1.74 |

Numbers for single GPU are obtained using the master branch.

Note: previously I was able to run 33b q4_0 with just the GTX 1070 on master; there may be something on master that has increased RAM usage since.

SlyEcho commented:
> 33b

You mean 30B? I can run 30B Q4_0 with my 8 GB card with 20 layers loaded only.

howard0su commented on llama.cpp:
```diff
@@ -1018,45 +1021,46 @@ static void llama_model_load_internal(
             std::string layers_i = "layers." + std::to_string(i);
 
-            layer.attention_norm = ml->get_tensor(layers_i + ".attention_norm.weight", {n_embd}, backend);
+            layer.attention_norm = ml->get_tensor(layers_i + ".attention_norm.weight", {n_embd}, i, backend);
```
we can decide which card we should put the layer on in this function (backend + device_id).

We should make backend like CPU = 0xFFFFFFFF; CUDA=0xEEEEEE00; (256 cards supported); etc.

JohannesGaessler replied:
This PR implements two ways to utilize multiple GPUs: tensors can either be put on a single GPU or they can be split across multiple GPUs. I think that tensors that require little computation should be put on a single GPU to reduce data copying. In the function `ggml_cuda_load_data` single GPU tensors are distributed to GPUs based on which layer they are in. I guess this distribution could also be done here but I don't think it would make a significant difference.

JohannesGaessler commented:
>You mean 30B? I can run 30B Q4_0 with my 8 GB card with 20 layers loaded only.

"30B" seems to be a typo by Meta that has become dominant. In the paper they talk about a "33B" model so that is the term that I'm using.

SlyEcho commented on ggml-cuda.cu:
```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
 }
 
-static void ggml_cuda_mul_mat_q_f32(const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst) {
+inline void ggml_cuda_op_mul(
```
```suggestion
static inline void ggml_cuda_op_mul(
```

SlyEcho commented on ggml.h:
```diff
@@ -410,6 +414,24 @@ extern "C" {
+    struct ggml_compute_params {
+        enum ggml_task_type type;
+
+        int ith, nth;
```
```suggestion
        // current thread, total threads
        int ith, nth;
```

slaren commented on ggml-cuda.cu:
```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
-                // wait for data
-                CUDA_CHECK(cudaStreamWaitEvent(cudaStream, cudaEvent, 0));
+    // asq = actual size quantized, asf = actual size float
+    size_t src0_asq[GGML_MAX_DEVICES];
+    memset(src0_asq, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t src0_asf[GGML_MAX_DEVICES];
+    memset(src0_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t src1_asf[GGML_MAX_DEVICES];
+    memset(src1_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
+    size_t dst_asf[GGML_MAX_DEVICES];
+    memset(dst_asf, 0, sizeof(size_t)*GGML_MAX_DEVICES);
```
Minor nit, is there any reason to use `memset` here instead of initializing to `= {0}`, similar to the previous declarations?

JohannesGaessler replied:
Would that actually initialize the whole array? I think to remember that in C the initializer `= {0}` will only initialize the first value; I only added the `= {nullptr}` for readability.

slaren replied:
Yes, this would initialize all the elements to zero ([ref](https://en.cppreference.com/w/c/language/array_initialization)).

slaren commented on ggml-cuda.cu:
```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
+                // copy dst to host if necessary
+                if (!dst_on_device) {
+                    // wait with memcpy until main stream is done
+                    CUDA_CHECK(cudaStreamWaitEvent(cudaStream_memcpy_dst, cudaEvent_main, 0));
```
Why not do this in the stream `cudaStream_main` instead? Is there any parallelization gained by doing this?

JohannesGaessler replied:
There does not seem to be a measurable performance difference. I added these because I think they will be useful for synchronization: you can wait for computation to finish by synchronizing the main streams and for data transfer to finish by waiting for the data copy streams. It wouldn't be a problem to remove them until they are actually useful though.

slaren replied:
If it is still not useful, I think it is better to remove it just to reduce the complexity of the code. I am not sure in what situations it may be desirable to wait for the computation instead of the copy, if you plan to do further computations on the GPU then the copy is not needed anyway.

SlyEcho commented on llama.cpp:
```diff
@@ -663,10 +663,10 @@ struct llama_model_loader {
     }
 
-    struct ggml_tensor * get_tensor_for(llama_load_tensor & lt, ggml_backend backend) {
+    struct ggml_tensor * get_tensor_for(llama_load_tensor & lt, int layer, ggml_backend backend) {
```
What if it were like this so the tensors that are not on a layer don't have to specify -1?
```suggestion
    struct ggml_tensor * get_tensor_for(llama_load_tensor & lt, ggml_backend backend, int layer = -1) {
```

slaren commented on ggml-cuda.cu:
```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
+    (void) i1;
+}
+
+template<enum ggml_cuda_op_type op_type, ggml_cuda_op_t op>
+static void ggml_cuda_op(const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst) {
```
I am not sure that there is a good reason to make this a template, the performance hit of doing these checks at runtime should be negligible. I would also consider replacing `ggml_cuda_op_type` with explicit checks for the type (from what I understand, the way this is used you could simply use `tensor->type`).

slaren replied:
It would also remove the need to use magic bit checks like this:
https://github.com/ggerganov/llama.cpp/blob/bee33f71759299591f85b7bc0d5622f1e64d724e/ggml-cuda.cu#L782

JohannesGaessler replied:
A template is a little more convenient for debugging since the NVIDIA compute sanitizer will then give you a better stacktrace but I would be fine with a regular function.

slaren replied:
With nsight systems it is possible to use NVTX to add markers, it may also work with the compute sanitizer, not sure.

JohannesGaessler replied:
How would changing `ggml_cuda_op` from a template help with the check to determine whether `src0` requires f32? The check uses `GGML_CUDA_OP_TYPE` which would still have to be passed as a parameter, right?

https://github.com/JohannesGaessler/llama.cpp/blob/f22982f84b86fd28f30f61d0c696d7d5f6efa507/ggml-cuda.cu#L36-L45

slaren replied:
Right, I wasn't sure if the `op_type` is needed at all. Still, I would suggest replacing `ggml_cuda_op_type` with explicit types to be more clear. Rather than an enum, this is really a struct with 3 booleans. More generally, you could replace it with a struct such as:
```C++
struct ggml_cuda_op_type {
    ggml_type src0_type, src1_type, dst_type;
}
```
Or just pass `src0_type` etc as individual parameters to the function, it doesn't matter. The goal here is to improve code clarity and remove possible sources of errors in the future.

JohannesGaessler replied:
For now I've just made it a boolean since only `src0` can be in need of quantization; it shouldn't be too hard to change if it becomes necessary.

ggerganov commented on ggml.h:
```diff
@@ -371,10 +373,12 @@ extern "C" {
 
         void * data;
+        void * data_device[GGML_MAX_DEVICES];
+        int i_device;
```
Consider moving all GPU related members into a struct in `ggml-cuda.h` which is referenced via:

```c
struct ggml_tensor {
    ...
    void * data_extra; // user-managed
    ...
};
```

For example:

```c
// ggml-cuda.h

struct ggml_tensor_gpu {
    int layer;
    void * device[GGML_CUDA_MAX_DEVICES];
    int idev;
};

// llama.cpp / ggml-cuda.cpp

...
struct ggml_tensor_gpu my_gpu_data;
tensor->data_extra = &my_gpu_data;
...
```

The extra structs will be managed by the user-code (i.e. `llama.cpp`) or by `ggml-cuda` / `ggml-opencl`.
The goal is to avoid coupling `ggml` core structs with LLM- / GPU- specific things, so that the interface remains more general-purpose.

Green-Sky replied:
What if NUMA nodes are considered different devices too?

JohannesGaessler replied:
Should I put the int representing the layer also into the GPU extra data struct or should it stay in `ggml_tensor`?

JohannesGaessler replied:
I decided to keep `layer` in `ggml_tensor`. Otherwise you would need to start adding extra data to `ggml_tensor` outside of `ggml-cuda.cu` (or extract it from the name).

ggerganov replied:
> What if NUMA nodes are considered different devices too?

Unlikely

> I decided to keep layer in ggml_tensor

The `layer` is specific to deep neural nets and it is not used in `ggml.c` as far as I can tell.
I think it should be moved out of `ggml_tensor`

JohannesGaessler replied:
I  also added a define `GGML_MAX_DEVICES` and a utility function `ggml_nbytes_split` to `ggml.h`. Should I move those as well?

ggerganov replied:
The `GGML_MAX_DEVICES` - yes

The `ggml_nbytes_split()` can go either way - as you prefer.

slaren commented:
I think `ggml_cuda_mul` and `ggml_cuda_mul_mat` can be removed from `ggml-cuda.h` now and made static.

JohannesGaessler commented:
I added a comment to explain the weird device to host memcpy for split tensors. Since I as the person to wrote the code won't know: are there other parts of the code that are unintuitive or difficult to understand?

JohannesGaessler commented:
I added a CLI argument that lets the user set the tensor split. On my system a less VRAM efficient split of 3:1 seems to do better than 2:1 because it's more efficient in terms of compute:

| Model    | GPU                               |     ms/t |      t/s |
|----------|-----------------------------------|----------|----------|
| 7b q4_0  | GTX 1070                          |    71.37 |    14.01 |
| 7b q4_0  | GTX 1070 + GTX 1050 ti, 2:1 split |    68.66 |    14.56 |
| 7b q4_0  | GTX 1070 + GTX 1050 ti, 3:1 split |    59.03 |    16.94 |
| 13b q4_0 | GTX 1070                          |   134.19 |     7.45 |
| 13b q4_0 | GTX 1070 + GTX 1050 ti, 2:1 split |   128.13 |     7.80 |
| 13b q4_0 | GTX 1070 + GTX 1050 ti, 3:1 split |   109.14 |     9.15 |
| 33b q4_0 | GTX 1070                          | Unusable | Unusable |
| 33b q4_0 | GTX 1070 + GTX 1050 ti, 2:1 split |   575.12 |     1.74 |
| 33b q4_0 | GTX 1070 + GTX 1050 ti, 3:1 split |   571.10 |     1.75 |

fgdfgfthgr-fox commented:
> The performance increase on my test systems is relatively low (+70% t/s when going from 1x GTX TITAN X to 4x GTX TITAN X).

May I ask the RAM and CPU you used for this test system?

JohannesGaessler commented:
It's a server set up by my institute. The CPU is a Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz. I don't know what kind of RAM they put in since I'm not aware of any way to query this without superuser privileges. The test is also complicated by the fact that other people were using the machine at the time (and still are).

JohannesGaessler commented:
Alright, unless I'm forgetting something this PR should now be ready to be merged from my end.

JohannesGaessler commented:
There seems to be an issue with f16 models.

JohannesGaessler commented:
I fixed f16. I should perhaps mention that this quantization type does **not** support multiple GPUs; I plan to work on better f16 support in the future (and see if that will allow you to use less VRAM) and will change it then.

JohannesGaessler commented:
I was using very short prompt for testing. There is an issue with long prompts.

huichen commented:
I'm getting following error:

`cuBLAS error 14 at ggml-cuda.cu:759`

when running

`./main -m ~/llms/ggml-vic13b-q5_1.bin  -p "hello" -ngl 1`

Followed your instruction in building the binaries.

JohannesGaessler commented:
Can you quickly check whether the code produces correct results with `--tensor-split 1,0,0,0`?

JohannesGaessler commented:
I fixed the issue with prompt processing. f16 still seems to have a bug somewhere with multiple GPUs.

JohannesGaessler commented:
I fixed the f16 issues. As a side effect f16 t/s also went up by ~100% because until now it was always using the general f16 f32 matrix multiplication function rather than the dequantization + matrix vector multiplication kernel that I implemented.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -440,21 +428,48 @@ static void ggml_cuda_pool_free(void * ptr, size_t size) {
+static int g_device_count = -1;
+static float g_tensor_split[GGML_CUDA_MAX_DEVICES] = {0};
+
 static cublasHandle_t g_cublasH = nullptr;
```
Thank you for pointing this out. Strangely enough, on my system with 2 GPUs my implementation is working even with only one cuBLAS handle (and it *is* being invoked from both devices). In any case, doing what the documentation says is probably for the best.

JohannesGaessler commented:
@huichen Can you do another test? The problem may have been caused by me only using one cuBLAS handle instead of one per GPU.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -514,330 +567,346 @@ static cudaError_t ggml_cuda_h2d_tensor_2d(void * dst, const struct ggml_tensor
+                    if (i0 == i0_offset_low) {
+                        i01_low = row_low % ne01;
+                    }
+                    if (i0 == i0_offset_high) {
+                        i01_high = row_high % ne01;
+                    }
+                }
+                const int64_t i01_diff = i01_high - i01_low;
```
`row_high > row_low` is always true. `row_high` is only modified when you're in the `i03`, `i02` bin where there is the boundary to the rows of the next GPU. If `row_low` is in the same bin, then its modulo must be smaller than the modulo of `row_high`. If `row_low` is in a previous bin, then it's 0. So `i01_diff` should always be positive.

I'm not particularly happy with the current state of the code in terms of readability; if you want to talk about it orgive feedback we can talk on my Mumble server or something.

JohannesGaessler replied:
>I am still trying to understand why we are converting a 3D(?) mat mul computation into multiple vector * vector operations.

Prompt processing is done in batches with `ne1 > 1` and `ne02 == ne12 == ne2 > 1`. I took over the design of the preexisting code and I think the point is to interleave memory transfers with computation so the latency is reduced. For token generation you have `ne03 == ne02 == 1` so you can essentially ignore those loops. You also have `ne1 == 1`, i.e. matrix vector multiplications.

JohannesGaessler commented on ggml-cuda.cu:
```diff
@@ -514,330 +567,346 @@ static cudaError_t ggml_cuda_h2d_tensor_2d(void * dst, const struct ggml_tensor
+    const uint64_t ne10 = src1->ne[0];
+    const uint64_t ne11 = src1->ne[1];
+
+    const uint64_t i0_diff = i0_high - i0_low;
```
`i0_high` is exclusive. In the simplest case where you have only one GPU you have `i0_low = 0` and `i0_high = ne01`. Why would the last row not be calculated?

huichen commented:
> @huichen Can you do another test? The problem may have been caused by me only using one cuBLAS handle instead of one per GPU.

It works now, both with and without --tensor-split. Following are eval times for different --tensor-split I experimented, numbers are averaged from three runs each (all layers offloaded to GPU)

|  --tensor-split   | prompt eval ms/t  | eval ms/t |
|  ----  | ----  | ---- |
| 1,0,0,0  | 9.17 | 65.80 |
| 1,1,0,0  | 8.84 | 49.90 |
| 1,1,1,0  | 9.08 | 48.18 |
| 1,1,1,1  | 9.29 | 43.85 |

Cheers!

cmp-nct commented:
@JohannesGaessler  Did you see that project ? https://github.com/turboderp/exllama/tree/master/exllama_ext/cuda_func
Looks like a ton of kernels on MIT license including full matmul for half and quantized variants, rope, norm etc.
Not sure where it comes from

JohannesGaessler commented:
Yes, as I've said a dozen times before: I have seen the exllama repository. But it's not as simple as copy-pasting code from one project to another. I already know what to implement, the problem is just doing it in a way that actually works. Most of my time doing development is spent hunting down ggml-specific bugs and the exllama repository does not help here. Also to get good performance you have to e.g. consider the memory layout of ggml tensors and adjust your implementation accordingly. So using exllama code 1:1 in ggml probably won't work well.

SlyEcho commented:
Getting good matmul perf is amazingly hard. CLBlast is like 2x slower than cuBLAS/rocBLAS for example. But these are libraries for general use, for llama.cpp it may be possible to create something that is custom but still very fast.

ggerganov commented:
Just a quick update - sorry for the delayed review. Really focused on the Metal branch. Hope to finish is in a few days and then will come back to more active reviews.

There are a few other important PRs that are also pending and I need to get familiar with them before merging

JohannesGaessler commented:
Don't worry, I'm patient. Currently I'm working on GPU acceleration for the remaining tensors. If I get a working version before this PR gets merged, should I just keep pushing to this PR or save it for a new one?

ggerganov commented:
> should I just keep pushing to this PR or save it for a new one?

The current PR is a good addition on its own. Probably better to have the new stuff in a separate PR

quarterturn commented:
It compiles, finds my cuda devices, but it does not use them:
```
ggml_init_cublas: found 2 CUDA devices:
  1. Tesla P40
  2. Tesla P40
llama.cpp: loading model from ./models/65B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 22016
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 65B
llama_model_load_internal: ggml ctx size =    0.18 MB
llama_model_load_internal: mem required  = 38610.46 MB (+ 5120.00 MB per state)
llama_model_load_internal: [cublas] offloading 0 layers to GPU
llama_model_load_internal: [cublas] total VRAM used: 0 MB
.
llama_init_from_file: kv self size  = 1280.00 MB

system_info: n_threads = 24 / 48 | AVX = 1 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
main: interactive mode on.
```

JohannesGaessler commented:
You need to set the --n-gpu-layers CLI argument to utilize the GPUs.

quarterturn commented:
> You need to set the --n-gpu-layers CLI argument to utilize the GPUs.

Is there a way to know how many layers the model needs at a certain bit depth? I think 65B needs 80, based on what I saw in text-generation-webui.

KerfuffleV2 commented:
> I think 65B needs 80, based on what I saw in text-generation-webui (turns out it is 80)

If you know they'll all fit, you should be able to set it to an absurdly high number like 10000. From what I've seen people say, it's not an error to set it higher than the number of layers in the model. Of course, if your GPU doesn't have enough VRAM it's going to die once it tries to load the data.

ThioJoe commented:
> Is there a way to know how many layers the model needs at a certain bit depth?

@quarterturn - Copying from my comment from here: https://github.com/ggerganov/llama.cpp/discussions/1438#discussioncomment-5894719

> To find the number of layers for a particular model, run the program normally using that model and look for something like:
> llama_model_load_internal: n_layer    = 32

Also like [Green-Sky](https://github.com/Green-Sky) mentions, you can apparently add 1 to whatever number it shows to also load the 'output layer' into VRAM too:
> I will let you in on a secret. If you use 33, the output layer is also put into vram. :)
> yes, this is underdocumented...

quarterturn commented:
> > I will let you in on a secret. If you use 33, the output layer is also put into vram. :)
> > yes, this is underdocumented...

Interesting. 81 layers works for 65B. What I'd like most is more output. It seems like it's fixed to 512 characters.

```
main: build = 617 (46d4278)
main: seed  = 1685819860
ggml_init_cublas: found 2 CUDA devices:
  1. Tesla P40
  2. Tesla P40
llama.cpp: loading model from ./models/65B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 22016
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 65B
llama_model_load_internal: ggml ctx size =    0.18 MB
llama_model_load_internal: mem required  = 3865.46 MB (+ 5120.00 MB per state)
llama_model_load_internal: [cublas] offloading 80 layers to GPU
llama_model_load_internal: [cublas] offloading output layer to GPU
llama_model_load_internal: [cublas] total VRAM used: 34745 MB
....................................................................................................
llama_init_from_file: kv self size  = 5120.00 MB

system_info: n_threads = 40 / 48 | AVX = 1 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
main: interactive mode on.
Reverse prompt: '### Instruction:

'
sampling: repeat_last_n = 64, repeat_penalty = 1.300000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.100000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = 128, n_keep = 21


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.

 Below is an instruction that describes a task. Write a response that appropriately completes the request.
> Create a study guide for Japanese particles. List them in rank of how commonly they are encountered. Make it a numbered list separated by returns. Give an example of usage for each particle.
1) は (wa): used to indicate the topic, or what is being talked about; can be translated as "about" when followed by verbs and adjectives in their -te form ("~tte")
2) が (ga): indicates that something belongs to someone/something else. Can also mean "when", depending on context:
3) を(wo)/お (o): used for direct objects
>
```
Is it possible to get it to generate more? I have definitely seen it generate more on the regular llama.cpp in CPU mode.

KerfuffleV2 commented:
> Is it possible to get it to generate more? 

Run the command with `--help`. There are commandline options to control context length and the number of tokens to generate. This doesn't have anything to do with running on the GPU.

quarterturn commented:
> > Is it possible to get it to generate more?
> 
> Run the command with `--help`. There are commandline options to control context length and the number of tokens to generate. This doesn't have anything to do with running on the GPU.

Is it batch-size? If it is, it is ignoring it:
(my gpt4all.sh )
```
#
# Temporary script - will be removed in the future
#

cd `dirname $0`
cd ..


./main --n-gpu-layers 81 --color --instruct --threads 40 \
       --model ./models/65B/ggml-model-q4_0.bin  \
       --file ./prompts/alpaca.txt \
       --ctx_size 2048 -n -1 \
       -b 1024 \
       --repeat_last_n 64 --repeat_penalty 1.3 \
       --n_predict 128 --temp 0.1 --top_k 40 --top_p 0.95
```

what it reports after loading the model:
```
sampling: repeat_last_n = 64, repeat_penalty = 1.300000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.100000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = 128, n_keep = 21

```

KerfuffleV2 commented:
-n N, --n-predict N   number of tokens to predict (default: -1, -1 = infinity)

Please create an issue or discussion topic in the Q&A section when you have general questions instead of asking in pull requests.

Comments in pull requests are primarily for discussing that specific pull.

JohannesGaessler commented:
Both CUDA and OpenCL seem to be working correctly in short tests. However, right now I'm too tired to test this PR in detail. I'll do it tomorrow morning.

KerfuffleV2 commented:
```
ggml_init_cublas: found 1 CUDA devices:
  1. NVIDIA GeForce GTX 1060 6GB
```

Works for me on Linux, performance seems the same as master. Identical results with the same seed also. Perplexity (and I'd assume just cuBLAS prompt evaluation in general) is noticeably slower though. (Output abbreviated below.)

### `master`

```
$ taskset -c 0-5 ./perplexity -f /path/wiki.test.raw -m /path/llama-7b.ggmlv3.q4_0.bin -t 6 -ngl 0
system_info: n_threads = 6 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 3.66 seconds per pass - ETA 39 minutes
[1]4.4545,[2]4.9402,[3]5.8278
```

***

```
$ taskset -c 0-5 ./perplexity -f /path/wiki.test.raw -m /path/llama-7b.ggmlv3.q4_0.bin -t 6 -ngl 33
system_info: n_threads = 6 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 3.37 seconds per pass - ETA 36 minutes
[1]4.4547,[2]4.9404
```


### this pull

```
$ taskset -c 0-5 ./perplexity -f /path/wiki.test.raw -m /path/llama-7b.ggmlv3.q4_0.bin -t 6 -ngl 0
system_info: n_threads = 6 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 5.47 seconds per pass - ETA 59 minutes
[1]4.4545,[2]4.9402
```

***

```
$ taskset -c 0-5 ./perplexity -f /path/wiki.test.raw -m /path/llama-7b.ggmlv3.q4_0.bin -t 6 -ngl 33
system_info: n_threads = 6 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
perplexity: calculating perplexity over 655 chunks, batch_size=512
perplexity: 5.12 seconds per pass - ETA 55 minutes
[1]4.4545,[2]4.9402
```

JohannesGaessler commented:
Thank you for pointing this out, I forgot to check it. The reason is most likely that on master there are some f16 f32 matrix multiplications where f32 is converted to f16 and then cuBLAS is used to do f16 matrix multiplication. In this PR however the f16 data is converted to f32 and then cuBLAS is used to do f32 matrix multiplication. Originally I did not want to touch the f16 f32 code at all but it caused issues for CPU layers in combination with >1 GPU. 

As a user I don't really care about prompt processing speed because it's so fast anyways. As a developer waiting longer for perplexity calculations could be an issue though. I would be fine with delaying the merging of this PR if it's desired to fix prompt processing speed beforehand; otherwise I will at a later point make another PR to fix it.

JohannesGaessler commented:
I forgot to mention: the f16 f32 matrix multiplications are mostly related to the KV cache and they're done on permuted matrices. To use regular cuBLAS matrix multiplication you first need to copy the matrices to make them contiguous but I'm currently trying to develop a kernel that operates directly on the permuted memory layout. Ideally that will then be faster than master anyways (for GPU layers).

KerfuffleV2 commented:
> As a user I don't really care about prompt processing speed because it's so fast anyways. As a developer waiting longer for perplexity calculations could be an issue though. 

This is just the opinion of a random person, but prompt processing is fast compared to inference for sure but at least on relatively old hardware something like an 800 token prompt still takes a pretty noticeable time to evaluate especially on larger models like 33b+. Roughly doubling that time from, I don't know, 15-20sec to 30-40sec is quite noticeable.

Also, with the advances to context size technology and some models being fine tuned to allow >2048 sized context it's likely in the next months or whatever longer prompts are going to get more common.

I've also been doing experimentation on various stuff and using the `perplexity` tool to check the effect so making it twice as slow (and it's already pretty low for big models) definitely will make me sad.

JohannesGaessler commented:
One workaround that might work would be to re-add the implementation on master and to only enable it when there is a single GPU.

JohannesGaessler committed:
Fixed single GPU performance regression

JohannesGaessler commented:
I pushed a fix that re-adds the old code for single GPU prompt processing. Prompt processing seems to still be slightly slower but only by ~20%. I'm not sure what's causing the remaining difference; investigating and optimizing the performance would probably take some time. In any case, @KerfuffleV2 , can you please check the performance on your system?

KerfuffleV2 commented:
Definitely helps.

### `-ngl 0`

master: 3.66 seconds per pass - ETA 39 minutes
before: 5.47 seconds per pass - ETA 59 minutes
now: 4.16 seconds per pass - ETA 45 minutes

### `-ngl 33`

master: 3.37 seconds per pass - ETA 36 minutes
before: 5.12 seconds per pass - ETA 55 minutes
now: 3.74 seconds per pass - ETA 40 minutes

JohannesGaessler commented:
Alright, my current status is this: I'm working on another PR based on this one that adds GPU acceleration for the entire LLaMa model rather than just matrix multiplication and component-wise multiplication. On my hardware this WIP version is already faster than master and this PR when it comes to token generation. I would like to get a working version for that first, and then optimize performance afterwards. Optimizing the performance of this PR now seems like it would be a bad investment of my time since I think I will be able to get better performance with a fundamentally different implementation anyways. So I think the two options for this PR are to either merge it and accept a performance regression for prompt processing or to wait until I have the next PR ready which I think will be faster than master across the board.

JohannesGaessler commented:
llama.cpp master currently reserves 1 GB of scratch buffer memory for 33b and smaller models. Translating this 1:1 to VRAM would be simple but inefficient because these scratch buffers are much larger than they need to be. Also you only need that much memory initially during prompt processing and there the amount required is proportional to the batch size. However, for low-end cards a large batch size probably has diminishing returns anyways. You also need some VRAM for the KV cache but I haven't worked out how much that will be exactly.

My current testing targets are an RTX 3090, a GTX 1070, and a GTX 1050 ti. If I can get universally better performance for all of these that's great, otherwise I plan to add something like a `--low-vram` option.

ggerganov commented on llama.cpp:
```diff
@@ -689,6 +695,17 @@ struct llama_model_loader {
+
+#ifdef GGML_USE_CUBLAS
+        if (backend == GGML_BACKEND_GPU || backend == GGML_BACKEND_GPU_SPLIT) {
+            struct ggml_tensor_extra_gpu * extra = new struct ggml_tensor_extra_gpu;
```
Have to deallocate these extra buffers at the end of the program

JohannesGaessler replied:
They are deallocated [here](https://github.com/JohannesGaessler/llama.cpp/blob/11af67866e927ddab9dd86735a7e140fcc93a253/llama.cpp#L203-L207) and [here](https://github.com/JohannesGaessler/llama.cpp/blob/11af67866e927ddab9dd86735a7e140fcc93a253/ggml-cuda.cu#L1141-L1158) when `llama_model` gets deconstructed

ggerganov commented:
> So I think the two options for this PR are to either merge it and accept a performance regression for prompt processing or to wait until I have the next PR ready which I think will be faster than master across the board.

The perplexity computation speed is quite important to me so if the fix is not trivial, I prefer to wait for the other PR that will hopefully not degrade the performance. The multi-GPU addition is really great, but from practical point of view, we need the perplexity speed right now, even more so because of #1684 efforts that are about to get merged.

I'll leave this PR open for now - if I get the time, I might try to fix the regression and merge it (if the other PR is not ready yet)

JohannesGaessler commented:
I did a quick test with my WIP PR. The perplexity speed is 2.23 ms/t vs. 3.07 ms/t on master. The PR has GPU acceleration for add, SiLU, RMS norm, cpy, reshape, view, transpose, and RoPE. As a third option I could prioritize getting that PR in a state that can be merged and then deliver kernels for the rest of the operations in a later PR.

ggerganov commented:
> As a third option I could prioritize getting that PR in a state that can be merged and then deliver kernels for the rest of the operations in a later PR.

Up to you - whichever way you prefer

JohannesGaessler commented:
I'll do it.

JohannesGaessler commented:
Obsolete due to the merging of https://github.com/ggerganov/llama.cpp/pull/1703
